File: step_09_stack_queues/9.3_monotonic_stack_queue/P004_number_of_nges_to_right/Solution.java
/**
 * Problem: Number of Next Greater Elements to the Right
 * Difficulty: MEDIUM | XP: 25
 *
 * Given an array of N integers, for each element find the count of
 * elements strictly greater than it that lie to its right.
 *
 * E.g., arr=[3,4,2,7,5,1] -> [3,2,2,0,0,0]
 *   arr[0]=3: {4,7,5} are greater -> count=3
 *   arr[1]=4: {7,5} -> count=2
 *   arr[2]=2: {7,5} -> count=2
 *   arr[3]=7: {} -> count=0
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE
    // Time: O(n^2)  |  Space: O(1) extra
    //
    // For each element at index i, scan all elements to its right
    // and count those strictly greater than arr[i].
    // ============================================================
    public static int[] bruteForce(int[] arr) {
        int n = arr.length;
        int[] result = new int[n];
        for (int i = 0; i < n; i++) {
            int count = 0;
            for (int j = i + 1; j < n; j++) {
                if (arr[j] > arr[i]) count++;
            }
            result[i] = count;
        }
        return result;
    }

    // ============================================================
    // APPROACH 2: OPTIMAL -- Modified Merge Sort
    // Time: O(n log n)  |  Space: O(n)
    //
    // During merge sort, when we merge left and right halves,
    // elements in the right half have higher original indices.
    // If a right-half element is greater than a left-half element,
    // it is a valid NGE for that left element. Count remaining right
    // elements at the moment a left element is placed.
    //
    // Key: All elements in right half originally appeared AFTER
    // all elements in left half, so they are valid "right" elements.
    // ============================================================
    static int[] resultGlobal;

    public static int[] optimal(int[] arr) {
        int n = arr.length;
        resultGlobal = new int[n];
        // Pair: {value, originalIndex}
        int[][] indexed = new int[n][2];
        for (int i = 0; i < n; i++) {
            indexed[i][0] = arr[i];
            indexed[i][1] = i;
        }
        mergeSort(indexed, 0, n - 1);
        return resultGlobal;
    }

    private static void mergeSort(int[][] arr, int l, int r) {
        if (l >= r) return;
        int mid = (l + r) / 2;
        mergeSort(arr, l, mid);
        mergeSort(arr, mid + 1, r);
        merge(arr, l, mid, r);
    }

    private static void merge(int[][] arr, int l, int mid, int r) {
        int n1 = mid - l + 1;
        int n2 = r - mid;
        int[][] left = new int[n1][2];
        int[][] right = new int[n2][2];
        System.arraycopy(arr, l, left, 0, n1);
        System.arraycopy(arr, mid + 1, right, 0, n2);

        int i = 0, j = 0, k = l;
        while (i < n1 && j < n2) {
            if (left[i][0] < right[j][0]) {
                // right[j] > left[i]: all of right[j..n2-1] are >= right[j] > left[i]
                // They all appeared to the right of left[i] originally.
                resultGlobal[left[i][1]] += (n2 - j);
                arr[k++] = left[i++];
            } else {
                arr[k++] = right[j++];
            }
        }
        while (i < n1) arr[k++] = left[i++];
        while (j < n2) arr[k++] = right[j++];
    }

    // ============================================================
    // APPROACH 3: BEST -- BIT (Binary Indexed Tree / Fenwick Tree)
    // Time: O(n log n)  |  Space: O(max_val)
    //
    // Process from right to left. Maintain a frequency BIT over
    // (compressed) values. For each arr[i], query BIT for count
    // of elements in (arr[i], max] that are already inserted.
    // Then insert arr[i] into the BIT.
    // ============================================================
    static int[] bit;
    static int bitSize;

    public static int[] best(int[] arr) {
        int n = arr.length;
        int[] result = new int[n];

        // Coordinate compression
        int[] sorted = arr.clone();
        Arrays.sort(sorted);
        // Remove duplicates and build rank map
        Map<Integer, Integer> rank = new HashMap<>();
        int r = 0;
        for (int v : sorted) {
            if (!rank.containsKey(v)) rank.put(v, ++r);
        }
        bitSize = r;
        bit = new int[bitSize + 2];

        for (int i = n - 1; i >= 0; i--) {
            int rk = rank.get(arr[i]);
            // Count of elements > arr[i] = total inserted - count of elements <= arr[i]
            int totalInserted = bitQuery(bitSize);
            int countLE = bitQuery(rk);
            result[i] = totalInserted - countLE;
            bitUpdate(rk);
        }
        return result;
    }

    private static void bitUpdate(int i) {
        for (; i <= bitSize; i += i & (-i)) bit[i]++;
    }

    private static int bitQuery(int i) {
        int s = 0;
        for (; i > 0; i -= i & (-i)) s += bit[i];
        return s;
    }

    // ============================================================
    // TESTING
    // ============================================================
    public static void main(String[] args) {
        System.out.println("=== Number of NGEs to Right ===\n");

        int[][][] tests = {
            {{3, 4, 2, 7, 5, 1}, {3, 2, 2, 0, 0, 0}},
            {{1, 2, 3, 4, 5},    {4, 3, 2, 1, 0}},
            {{5, 4, 3, 2, 1},    {0, 0, 0, 0, 0}},
            {{1},                {0}},
            {{3, 3, 3},          {0, 0, 0}},
            {{1, 3, 2, 4},       {3, 1, 1, 0}},
        };

        for (int[][] test : tests) {
            int[] arr = test[0];
            int[] expected = test[1];

            int[] b = bruteForce(arr.clone());
            int[] o = optimal(arr.clone());
            int[] r = best(arr.clone());

            boolean pass = Arrays.equals(b, expected)
                        && Arrays.equals(o, expected)
                        && Arrays.equals(r, expected);

            System.out.printf("arr=%s%n", Arrays.toString(arr));
            System.out.printf("  Expected: %s%n", Arrays.toString(expected));
            System.out.printf("  Brute:    %s%n", Arrays.toString(b));
            System.out.printf("  Optimal:  %s%n", Arrays.toString(o));
            System.out.printf("  Best:     %s  [%s]%n%n", Arrays.toString(r), pass ? "PASS" : "FAIL");
        }
    }
}

File: step_09_stack_queues/9.3_monotonic_stack_queue/P004_number_of_nges_to_right/solution.py
"""
Problem: Number of Next Greater Elements to the Right
Difficulty: MEDIUM | XP: 25

Given an array arr[] of N integers, for each element find the count of
elements strictly greater than it that lie to its right.

E.g., arr = [3, 4, 2, 7, 5, 1]
  idx 0 (val=3): elements to right greater than 3 -> {4,7,5} -> count=3
  idx 1 (val=4): {7,5} -> count=2
  idx 2 (val=2): {7,5} -> count=2
  idx 3 (val=7): {} -> count=0
  idx 4 (val=5): {} -> count=0
  idx 5 (val=1): {} -> count=0
Output: [3, 2, 2, 0, 0, 0]
"""
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE
# Time: O(n^2)  |  Space: O(1) extra
#
# For each element, scan all elements to its right and count
# those strictly greater than it.
# ============================================================
def brute_force(arr: List[int]) -> List[int]:
    """Nested loops: for each i, count j > i where arr[j] > arr[i]."""
    n = len(arr)
    result = [0] * n
    for i in range(n):
        count = 0
        for j in range(i + 1, n):
            if arr[j] > arr[i]:
                count += 1
        result[i] = count
    return result


# ============================================================
# APPROACH 2: OPTIMAL -- Merge Sort Based Count
# Time: O(n log n)  |  Space: O(n)
#
# Use a modified merge sort. During the merge step, when we take
# an element from the right half (it's greater than elements in
# the left half that haven't been merged yet), increment their counts.
#
# We augment each element with its original index so we can update
# the correct position in the result array.
# ============================================================
def optimal(arr: List[int]) -> List[int]:
    """Modified merge sort: count NGEs using the merge step."""
    n = len(arr)
    result = [0] * n
    # Pair each element with its index: (value, original_index)
    indexed = [(arr[i], i) for i in range(n)]

    def merge_sort(lst: List) -> List:
        if len(lst) <= 1:
            return lst
        mid = len(lst) // 2
        left = merge_sort(lst[:mid])
        right = merge_sort(lst[mid:])
        return merge(left, right)

    def merge(left: List, right: List) -> List:
        merged = []
        i = j = 0
        while i < len(left) and j < len(right):
            if left[i][0] < right[j][0]:
                # right[j] is greater than left[i]
                # All remaining elements in right are also >= right[j],
                # but we want strictly greater, so count elements in right
                # that are strictly > left[i].
                # However in merge sort we count in sorted order, which
                # doesn't directly give us "elements to the right in original array".
                # We need a different formulation.
                merged.append(left[i])
                i += 1
            else:
                merged.append(right[j])
                j += 1
        merged.extend(left[i:])
        merged.extend(right[j:])
        return merged

    # The merge-sort approach for "count of greater elements to the right"
    # requires careful tracking. We count: for each element in the left half
    # during merge, how many elements from the right half are strictly greater.
    # Since right half has higher original indices (elements to the right),
    # each right element that passes over a left element is a valid NGE.
    #
    # Correct implementation: count elements from right half that merge before
    # each left element (meaning they are greater).
    result_arr = [0] * n
    indexed2 = [(arr[i], i) for i in range(n)]

    def merge_sort2(lst):
        if len(lst) <= 1:
            return lst[:]
        mid = len(lst) // 2
        left = merge_sort2(lst[:mid])
        right = merge_sort2(lst[mid:])
        return merge2(left, right)

    def merge2(left, right):
        merged = []
        i = j = 0
        # Count how many from right have been placed before each left element
        right_count = 0
        while i < len(left) and j < len(right):
            if left[i][0] < right[j][0]:
                # right[j] > left[i]: all of right[j:] that come before left[i]
                # are NGEs of left[i]. But we want only those from original right.
                # In merge sort, all right half elements have original indices > all left half elements.
                # Count of right elements placed so far that came BEFORE left[i] = right_count.
                # But we also need remaining right elements > left[i].
                # Actually: len(right) - j counts remaining right elements > left[i]
                # (since sorted order: right[j] <= right[j+1] <= ...)
                # Wait - we want strictly greater. Since right[j] > left[i],
                # remaining right[j:] might not all be > left[i].
                # This approach counts elements in right half that are greater,
                # but those are elements with HIGHER original positions.
                result_arr[left[i][1]] += len(right) - j
                merged.append(left[i])
                i += 1
            else:
                merged.append(right[j])
                j += 1
        while i < len(left):
            # No more right elements; remaining right that are greater = 0 already counted
            merged.append(left[i])
            i += 1
        while j < len(right):
            merged.append(right[j])
            j += 1
        return merged

    merge_sort2(indexed2)
    return result_arr


# ============================================================
# APPROACH 3: BEST -- BIT (Binary Indexed Tree / Fenwick Tree)
# Time: O(n log n)  |  Space: O(max_val)
#
# Process from right to left. For each element, query the BIT
# for count of elements already inserted that are strictly greater.
# Then insert the current element into the BIT.
# This works for discrete values (compress coordinates if needed).
# ============================================================
def best(arr: List[int]) -> List[int]:
    """BIT-based right-to-left processing: query prefix sum for count of NGEs."""
    n = len(arr)
    result = [0] * n

    # Coordinate compression
    sorted_unique = sorted(set(arr))
    rank = {v: i + 1 for i, v in enumerate(sorted_unique)}  # 1-indexed
    max_rank = len(sorted_unique)

    bit = [0] * (max_rank + 2)

    def update(i: int) -> None:
        while i <= max_rank:
            bit[i] += 1
            i += i & (-i)

    def query(i: int) -> int:
        """Sum of frequencies in [1..i]."""
        s = 0
        while i > 0:
            s += bit[i]
            i -= i & (-i)
        return s

    # Process right to left
    for i in range(n - 1, -1, -1):
        r = rank[arr[i]]
        # Count elements greater than arr[i] = total inserted - count of elements <= arr[i]
        total_inserted = query(max_rank)
        count_le = query(r)
        result[i] = total_inserted - count_le
        update(r)

    return result


# ============================================================
# TESTING
# ============================================================
if __name__ == "__main__":
    print("=== Number of NGEs to Right ===\n")

    tests = [
        ([3, 4, 2, 7, 5, 1],     [3, 2, 2, 0, 0, 0]),
        ([1, 2, 3, 4, 5],         [4, 3, 2, 1, 0]),
        ([5, 4, 3, 2, 1],         [0, 0, 0, 0, 0]),
        ([1],                     [0]),
        ([3, 3, 3],               [0, 0, 0]),
        ([1, 3, 2, 4],            [3, 1, 1, 0]),
    ]

    for arr, expected in tests:
        b = brute_force(arr[:])
        o = optimal(arr[:])
        r = best(arr[:])
        status = "PASS" if b == o == r == expected else "FAIL"
        print(f"arr={arr}")
        print(f"  Expected: {expected}")
        print(f"  Brute:    {b}")
        print(f"  Optimal:  {o}")
        print(f"  Best:     {r}  [{status}]")
        print()
